File: leetcode_543_Diameter_of_Binary_Tree.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
   int val;
   TreeNode *left;
   TreeNode *right;
   TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int m = 0;
    int bt(TreeNode* root) {
        if (root == NULL)
            return 0;
        int l = bt(root->left);
        int r = bt(root->right);
        m = max(m , (l + r));
        return max(l, r) + 1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        bt(root);
        return m;
    }
};

File: leetcode_543_Diameter_of_Binary_Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.m = 0
        def bt(root):
            if root == None:
                return 0
            l = bt(root.left)
            r = bt(root.right)
            self.m = max(self.m, (l + r))
            return max(l,r) + 1
        bt(root)
        return self.m